gtsfm/frontend/detector_descriptor/detector_descriptor_base.py

```diff
@@ -43,9 +43,26 @@ def detect_and_describe(self, image: Image) -> Tuple[Keypoints, np.ndarray]:
             Corr. descriptors, of shape (N, D) where D is the dimension of each descriptor.
         """
 
+    def filter_by_response(self, keypoints: Keypoints, descriptors: np.ndarray) -> Tuple[Keypoints, np.ndarray]:
+        """Filter features according to their responses."""
+        if keypoints.responses is None:
+            return keypoints, descriptors
+
+        # Sort by responses.
+        sort_idxs = np.argsort(-keypoints.responses)
+        sort_idxs = sort_idxs[: self.max_keypoints]
+
+        return keypoints.extract_indices(sort_idxs), descriptors[sort_idxs]
+
+    @staticmethod
+    def filter_by_mask(mask: np.ndarray, keypoints: Keypoints, descriptors: np.ndarray) -> Tuple[Keypoints, np.ndarray]:
+        """Filter features with respect to a binary mask of the image."""
+        valid_idxs = np.flatnonzero([mask[round(v), round(u)] == 1 for (u, v) in keypoints.coordinates])
```

Comment:
thanks for the PR. Any reason not to just use `np.round(keypoints.coordinates)`?

Reply:
seems could be vectorized with no for loop in that case

Reply:
Good idea.

Reply:
This should technically be a part of `describe` operation? For example the opencv sift implementation takes in the mask too. It will try to find more points in the non-masked region. Post-detection filtering with a mask will be suboptimal in that case.

Reply:
Yes, I believe you're right about SIFT. I will use their built-in function.

However, data-driven feature extractors typically return all features output by the network that are above some confidence (set by the user), as the hidden layers cannot be manipulated at test time. The output features are then post-processed (usually just taking the top k features wrt their scores). I'm unaware of any data-driven pipelines that take in a mask as input and change their output with respect to this mask.

Reply:
@ayushbaid could you add a pointer/reference to where in the SIFT code they are looking for additional keypoint during `describe()`?

I think it would be cleaner to put this masking in just a single place, and not spread it out across detect/describe.

Reply:
changed to using `np.round(keypoints.coordinates)` instead of for-loop.

Reply:
@johnwlambert It seems as though OpenCV masks the _input_ before computing features as opposed to masking the output. I'm not sure if this is any different than computing _all_ features and then masking them based on their keypoint, but it probably makes more sense to just use OpenCV's since it's built into the `detectAndCompute` function.

Reply:
Now passing mask to `detectAndCompute`

Reply:
Adding link to OpenCV source where masking is executed for future reference: https://github.com/opencv/opencv/blob/4.x/modules/features2d/src/orb.cpp